lockfile/src/cyclonedx.rs

```diff
@@ -0,0 +1,313 @@
+fn from_purl<T: Component>(component: &T) -> anyhow::Result<Package> {
+    let purl_str = component
+        .purl()
+        .ok_or_else(|| anyhow!("Missing purl for {}:{}", component.name(), component.version()))?;
+    let purl = GenericPurl::<String>::from_str(purl_str)?;
+    let package_type = PackageType::from_str(purl.package_type()).map_err(|_| UnknownEcosystem)?;
+
+    // Determine the package name based on its type and potential namespace.
+    let name = match (package_type, purl.namespace()) {
+        (PackageType::Maven, Some(ns)) => format!("{}:{}", ns, purl.name()),
+        (PackageType::Npm | PackageType::Golang, Some(ns)) => format!("{}/{}", ns, purl.name()),
+        _ => purl.name().into(),
+    };
+
+    // Extract the package version
+    let pkg_version = purl
+        .version()
+        .ok_or(&component.version())
+        .map_err(|_| anyhow!("No version found for `{}`", name))?;
+
+    // Use the qualifiers from the PURL to determine the version details.
+    let version = purl
+        .qualifiers()
+        .iter()
+        .find_map(|(key, value)| match key.as_ref() {
+            "repository_url" => Some(PackageVersion::ThirdParty(ThirdPartyVersion {
+                version: pkg_version.into(),
+                registry: value.to_string(),
+            })),
+            "download_url" => Some(PackageVersion::DownloadUrl(value.to_string())),
+            "vcs_url" => {
+                if value.starts_with("git+") {
+                    Some(PackageVersion::Git(value.to_string()))
+                } else {
+                    None
+                }
+            },
+            _ => None,
+        })
+        .unwrap_or(PackageVersion::FirstParty(pkg_version.into()));
+
+    Ok(Package { name, version, package_type })
+}
+
+pub struct CycloneDX;
+
+impl Parse for CycloneDX {
+    fn parse(&self, data: &str) -> anyhow::Result<Vec<Package>> {
+        if let Ok(lock) = serde_json::from_str::<serde_json::Value>(data) {
+            let parsed: Bom<Vec<JsonComponent>> = serde_json::from_value(lock)?;
+            parsed.components.map_or(Ok(vec![]), |comp| {
+                let component_iter = filter_components(&comp);
+                component_iter.map(from_purl).collect()
+            })
+        } else {
+            let parsed: Bom<Components<XmlComponent>> = serde_xml_rs::from_str(data)?;
+            parsed.components.map_or(Ok(vec![]), |comp| {
+                let component_iter = filter_components(&comp.components);
+                component_iter.map(from_purl).collect()
+            })
+        }
+    }
+
+    fn is_path_lockfile(&self, path: &Path) -> bool {
+        path.file_name() == Some(OsStr::new("bom.json"))
+            || path.file_name() == Some(OsStr::new("bom.xml"))
+    }
+
+    fn is_path_manifest(&self, _path: &Path) -> bool {
+        false
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+
+    #[test]
+    fn parse_cyclonedx_1_5_json() {
```

Comment:
Is there a specific reason why the json in this test is inlined rather than using a fixture?

Reply:
It was more that I couldn't find a good `v1.5` example. All tools seems to still generate `v1.4`. Granted `v1.5` was released fairly recently. The fields used for parsing packages are the same.

I can just move that example to a file to make it more consistent.

Reply:
Sounds good.